File: frontend/app/src/lib/furigana.test.ts
import { parseFurigana } from "./furigana.ts";

/** What the pane ends up showing — and what a user copies out of it. */
const plainText = (annotated: string) =>
  parseFurigana(annotated)
    .map((segment) => segment.text)
    .join("");

describe("parseFurigana", () => {
  it("returns one plain run when there is nothing annotated", () => {
    expect(parseFurigana("こんにちは")).toEqual([{ text: "こんにちは" }]);
  });

  it("returns nothing for an empty string", () => {
    expect(parseFurigana("")).toEqual([]);
  });

  it("puts a reading over the maximal run of kanji before it", () => {
    expect(parseFurigana("天気《てんき》です")).toEqual([{ text: "天気", reading: "てんき" }, { text: "です" }]);
  });

  it("reads a realistic sentence with several annotations", () => {
    expect(parseFurigana("今日《きょう》は良《よ》い天気《てんき》ですね")).toEqual([
      { text: "今日", reading: "きょう" },
      { text: "は" },
      { text: "良", reading: "よ" },
      { text: "い" },
      { text: "天気", reading: "てんき" },
      { text: "ですね" },
    ]);
  });

  it("keeps the iteration and abbreviation marks inside the base", () => {
    expect(parseFurigana("人々《ひとびと》")).toEqual([{ text: "人々", reading: "ひとびと" }]);
    expect(parseFurigana("三ヶ月《さんかげつ》")).toEqual([{ text: "三ヶ月", reading: "さんかげつ" }]);
    expect(parseFurigana("〆切《しめきり》")).toEqual([{ text: "〆切", reading: "しめきり" }]);
  });

  it("annotates a CJK Extension A character", () => {
    expect(parseFurigana("㐁《てん》")).toEqual([{ text: "㐁", reading: "てん" }]);
  });

  it("keeps a compatibility ideograph inside the base", () => {
    // 﨑 is U+FA11, in the compatibility block rather than the unified one, and ordinary in a
    // name: left out of the kanji class the run stops short and みやざき lands over 﨑 alone.
    expect(parseFurigana("宮﨑《みやざき》")).toEqual([{ text: "宮﨑", reading: "みやざき" }]);
  });

  it("keeps an astral extension character inside the base", () => {
    // 𠮟 is U+20B9F (Extension B): one character, two UTF-16 units. Left out of the class the
    // run starts at 責 and しっせき would be written over 責 by itself.
    expect(parseFurigana("𠮟責《しっせき》")).toEqual([{ text: "𠮟責", reading: "しっせき" }]);
  });

  it("keeps 〇 and the small ヵ inside the base, like 々 and ヶ", () => {
    expect(parseFurigana("〇〇《まるまる》")).toEqual([{ text: "〇〇", reading: "まるまる" }]);
    expect(parseFurigana("一ヵ月《いっかげつ》")).toEqual([{ text: "一ヵ月", reading: "いっかげつ" }]);
  });

  it("keeps the full-size ケ/カ/ノ/ツ that spell a kanji word inside the base", () => {
    // Place and counter names write these as full-size katakana — 霞ケ関 is how the Tokyo Metro
    // spells Kasumigaseki — and they sit *inside* the word. Left out of the run the reading lands
    // over the tail alone and the pane teaches that 関 reads かすみがせき, which is the exact lie the
    // module exists to prevent; the annotateTranslation round-trip can't catch it, because a
    // partial base is still non-empty and still strips back to the translation.
    expect(parseFurigana("霞ケ関《かすみがせき》")).toEqual([{ text: "霞ケ関", reading: "かすみがせき" }]);
    expect(parseFurigana("四ツ谷《よつや》")).toEqual([{ text: "四ツ谷", reading: "よつや" }]);
    expect(parseFurigana("一ノ瀬《いちのせ》")).toEqual([{ text: "一ノ瀬", reading: "いちのせ" }]);
    expect(parseFurigana("一ケ月《いっかげつ》")).toEqual([{ text: "一ケ月", reading: "いっかげつ" }]);
    expect(parseFurigana("一カ所《いっかしょ》")).toEqual([{ text: "一カ所", reading: "いっかしょ" }]);
  });

  it("will not start a base at one of those letters, so a katakana word keeps them", () => {
    // The price of admitting ケ/カ/ノ/ツ: each is an ordinary katakana letter too. They are taken
    // only with a true kanji on both sides, so the run here is 水 and バケツ stays plain text —
    // みず written over ケツ水 would be the same lie in the other direction.
    expect(parseFurigana("バケツ水《みず》")).toEqual([{ text: "バケツ" }, { text: "水", reading: "みず" }]);
    expect(parseFurigana("ノート《のーと》")).toEqual([{ text: "ノート" }]);
    expect(parseFurigana("カタカナ語《ご》")).toEqual([{ text: "カタカナ" }, { text: "語", reading: "ご" }]);
  });

  it("trims a run back rather than ending it on one of those letters", () => {
    // A run has to end in a true kanji, and the base has to be the text immediately before the
    // group, so there is nothing left for かすみが to sit over: dropped, like any other reading
    // that doesn't follow kanji.
    expect(parseFurigana("霞ケ《かすみが》")).toEqual([{ text: "霞ケ" }]);
  });

  it("drops a reading that doesn't follow kanji rather than writing it over the wrong base", () => {
    // The backend guarantees only that stripping the groups reproduces the translation, never
    // where a group sits. Over the character before it, おねがい would be written over い alone —
    // wrong, in the pane someone is reading the Japanese out of, where a gap is merely a gap.
    expect(parseFurigana("お願い《おねがい》します")).toEqual([{ text: "お願いします" }]);
    expect(parseFurigana("[JA]《ジェイエー》 Hello")).toEqual([{ text: "[JA] Hello" }]);
  });

  it("drops a reading with nothing usable in front of it", () => {
    expect(parseFurigana("《よみ》です")).toEqual([{ text: "です" }]);
    expect(parseFurigana("今日 《きょう》")).toEqual([{ text: "今日 " }]);
    expect(parseFurigana("行《い》《く》")).toEqual([{ text: "行", reading: "い" }]);
  });

  it("drops an empty reading rather than emitting a ruby over nothing", () => {
    expect(parseFurigana("漢字《》を書く")).toEqual([{ text: "漢字を書く" }]);
  });

  it("merges the plain runs on either side of a dropped group", () => {
    expect(parseFurigana("ab《》cd")).toEqual([{ text: "abcd" }]);
  });

  it("preserves newlines and every other character exactly", () => {
    expect(parseFurigana("一行目《いちぎょうめ》\n\n二行目《にぎょうめ》！")).toEqual([
      { text: "一行目", reading: "いちぎょうめ" },
      { text: "\n\n" },
      { text: "二行目", reading: "にぎょうめ" },
      { text: "！" },
    ]);
  });

  it("leaves a lone bracket alone: only a whole group is markup", () => {
    expect(parseFurigana("《 is a bracket")).toEqual([{ text: "《 is a bracket" }]);
  });

  it("strips back to the plain translation, which is what the user copies", () => {
    expect(plainText("漢字《かんじ》を書《か》く")).toBe("漢字を書く");
    expect(plainText("[JA]《ジェイエー》 Hello")).toBe("[JA] Hello");
  });
});

File: backend/test/services/translation/furigana_test.rb
# frozen_string_literal: true

require "test_helper"

# The annotation rules, asked of the string itself. Result.for_request is what asks them of a
# translator's reply (result_test.rb); this is where each rule's own case lives.
class Translation::FuriganaTest < ActiveSupport::TestCase
  test "the translation with a reading after every run of kanji is what an annotation is" do
    text = "これは野球のバットですか？"
    furigana = "これは野球《やきゅう》のバットですか？"

    assert_equal furigana, Translation::Furigana.checked(furigana, text)
    assert_nil Translation::Furigana.rejection(furigana, text)
  end

  test "a string that is not one, or carries no readings at all, is nothing to reject" do
    # The prompt asks for an empty string when there is nothing to annotate, and structured
    # outputs can still hand back a number or a null. Neither is a furigana and neither is a
    # reading Claude tried to give the reader, so neither is worth a warning in the log.
    [ nil, 42, "", "これは野球のバットですか？" ].each do |value|
      assert_nil Translation::Furigana.checked(value, "これは野球のバットですか？"), value.inspect
      assert_nil Translation::Furigana.rejection(value, "これは野球のバットですか？"), value.inspect
    end
  end

  test "a translation with 《…》 of its own cannot be annotated, and the log says why" do
    # U+300A/U+300B are ordinary Japanese punctuation — book titles, emphasis — and the prompt
    # asks Claude to keep the original's punctuation style, so `He recommended 《Kokoro》 to me.`
    # translates to a sentence carrying a pair of them. The notation has no way to hold that: the
    # browser deletes every 《…》 group it finds, so the title itself would disappear from the
    # text the reader copies. The readings give way instead — and the reason is named, because
    # "it does not strip back" reads as a model defect where this is punctuation nobody can
    # annotate. Result#readings_omitted is what tells the reader (result_test.rb).
    text = "彼は《こころ》を勧めてくれた。"
    furigana = "彼《かれ》は《こころ》を勧《すす》めてくれた。"

    assert_nil Translation::Furigana.checked(furigana, text)
    assert_equal "the translation contains 《 or 》 of its own", Translation::Furigana.rejection(furigana, text)
  end

  test "an annotation that leaves a run of kanji bare is rejected whole" do
    # Annotating only the unfamiliar part is ordinary furigana convention on paper, and here it
    # is a wrong reading nothing downstream can catch: the notation records no base length, so
    # the browser can only read はいえん as the reading of the run in front of it — 新型肺炎, four
    # characters it was never written for. Nothing in the string is malformed, and the round trip
    # passes. Only the translation, which this side has and the browser doesn't, shows 記事 left
    # bare. So the whole string goes, and the reader is told there are no readings rather than
    # shown one that is a lie.
    text = "新型肺炎の記事"
    furigana = "新型肺炎《はいえん》の記事"

    assert_equal "the readings do not sit one per run of kanji", Translation::Furigana.rejection(furigana, text)
    assert_nil Translation::Furigana.checked(furigana, text)
  end

  test "a reading that sits after anything but a run of kanji is rejected whole" do
    # お願い《おねがい》 is the same defect from the other end: the group follows the い, so the
    # browser has no run to put おねがい over and drops that reading — while 願 stands there
    # unannotated. The browser dropping one reading quietly is exactly the silence this rule
    # exists to break.
    text = "お願いします"
    furigana = "お願い《おねがい》します"

    assert_equal "the readings do not sit one per run of kanji", Translation::Furigana.rejection(furigana, text)
  end

  test "a reading written for part of the run it follows is accepted, because nothing can measure it" do
    # The limit of the rule above, stated so nobody reads more into it than it says. 毎日東京 is
    # one maximal run, so 毎日東京《とうきょう》 has its group in the only place a group for that
    # run can sit; both sides agree on the base, and neither can know とうきょう was meant for
    # 東京 alone. Nothing but the prompt can ask for the whole run's reading, and it does, in so
    # many words (Prompt::SYSTEM).
    text = "毎日東京へ行きます"
    furigana = "毎日東京《とうきょう》へ行《い》きます"

    assert_equal furigana, Translation::Furigana.checked(furigana, text)
  end

  test "an annotation that is not the translation over again is rejected" do
    text = "これは野球のバットですか？"
    furigana = "これは野球《やきゅう》の secret バットですか？"

    assert_equal "it does not strip back to the translation", Translation::Furigana.rejection(furigana, text)
  end

  test "a reading holding an opening bracket is not one group, because the browser says it isn't" do
    # The browser's READING_GROUP allows neither bracket inside a reading, so it reads this as the
    # single group 《じ》 with 《かん before it. A Ruby pattern that allowed an opening bracket
    # there swallowed 《かん《じ》 whole, got 漢字 back and shipped furigana the browser then threw
    # away entirely — every reading in the reply lost, silently.
    assert_nil Translation::Furigana.checked("漢字《かん《じ》", "漢字")
  end

  test "a group with nothing in it annotates nothing, so it is rejected too" do
    # The browser renders no ruby for an empty reading, which would leave 漢字 bare while the
    # string looks annotated.
    assert_equal "a reading is empty", Translation::Furigana.rejection("漢字《》を書《か》く", "漢字を書く")
  end

  test "a run is what the browser says a run is, at every boundary it has a rule for" do
    # These are the cases frontend/app/src/lib/furigana.ts names one by one, each annotated the
    # way the browser would read it. Ruby and TypeScript decide independently where a run starts
    # and ends, and a disagreement here is a furigana this side accepts and the browser drops, or
    # the other way about — so every boundary the browser documents is exercised against the same
    # example.
    {
      # Full-size ケ, ノ, ツ and small ヶ inside a run: 霞ケ関, 一ノ瀬, 四ツ谷, 一ヶ月.
      "霞ケ関に行く" => "霞ケ関《かすみがせき》に行《い》く",
      "一ノ瀬さん" => "一ノ瀬《いちのせ》さん",
      "四ツ谷まで" => "四ツ谷《よつや》まで",
      "一ヶ月です" => "一ヶ月《いっかげつ》です",
      # The same letters as ordinary katakana: バケツ水 is the run 水, never ケツ水.
      "バケツ水" => "バケツ水《みず》",
      # A run may not end on one either, so 日本語ケーキ is the run 日本語.
      "日本語ケーキ" => "日本語《にほんご》ケーキ",
      # The marks that appear only in a kanji word: 〇, 々 and 〆.
      "〇〇株式会社" => "〇〇株式会社《まるまるかぶしきがいしゃ》",
      "人々が集まる" => "人々《ひとびと》が集《あつ》まる",
      # A compatibility ideograph (the 﨑 of 宮﨑) and an astral extension (𠮟).
      "宮﨑さん" => "宮﨑《みやざき》さん",
      "𠮟る" => "𠮟《しか》る"
    }.each do |text, furigana|
      assert_equal furigana, Translation::Furigana.checked(furigana, text), text
    end
  end
end
